Wrappers/Python/cil/optimisation/functions/ApproximateGradientSumFunction.py

```diff
@@ -0,0 +1,213 @@
+
+
+
+    Note
+    ----
+    The :meth:`~ApproximateGradientSumFunction.gradient` returns the approximate gradient depending on an index provided by the  :code:`sampler` method. 
+
+    Example
+    -------
+    Consider the objective is to minimise: 
+
+    .. math:: \sum_{i=1}^{n} F_{i}(x) = \sum_{i=1}^{n}\|A_{i} x - b_{i}\|^{2}
+
+    >>> list_of_functions = [LeastSquares(Ai, b=bi)] for Ai,bi in zip(A_subsets, b_subsets))   
+    >>> f = ApproximateGradientSumFunction(list_of_functions) 
+
+    >>> list_of_functions = [LeastSquares(Ai, b=bi)] for Ai,bi in zip(A_subsets, b_subsets)) 
+    >>> sampler = RandomSampling.random_shuffle(len(list_of_functions))
+    >>> f = ApproximateGradientSumFunction(list_of_functions, sampler=sampler)   
+
+
+    """
+
+    def __init__(self, functions, sampler=None):
+
+        if sampler is None:
+            sampler = Sampler.random_with_replacement(len(functions))
+
+        if not isinstance(functions, list):
+            raise TypeError("Input to functions should be a list of functions")
+        if not hasattr(sampler, "next"):
+            raise ValueError('The provided sampler must have a `next` method')
+
+        self.sampler = sampler
+
+        self._partition_weights = [1 / len(functions)] * len(functions)
+
+        self._data_passes_indices = []
+
+        super(ApproximateGradientSumFunction, self).__init__(*functions)
+
+    def __call__(self, x):
+        r"""Returns the value of the sum of functions at :math:`x`.
+
+        .. math:: (F_{1} + F_{2} + ... + F_{n})(x) = F_{1}(x) + F_{2}(x) + ... + F_{n}(x)
+
+        Parameters
+        ----------
+        x : DataContainer
+
+        --------
+        float
+            the value of the SumFunction at x
+
+
+        """
+        return super(ApproximateGradientSumFunction, self).__call__(x)
+
+    def full_gradient(self, x, out=None):
+        r"""Returns the value of the  full gradient of the sum of functions at :math:`x`.
+
+        .. math:: \nabla_x(F_{1} + F_{2} + ... + F_{n})(x) = \nabla_xF_{1}(x) + \nabla_xF_{2}(x) + ... + \nabla_xF_{n}(x)
+
+        Parameters
+        ----------
+        x : DataContainer
+        out: return DataContainer, if `None` a new DataContainer is returned, default `None`.
+
+        Returns
+        --------
+        DataContainer
+            the value of the gradient of the sum function at x or nothing if `out`  
+        """
+
+        return super(ApproximateGradientSumFunction, self).gradient(x, out=out)
+
+    @abstractmethod
+    def approximate_gradient(self, x, function_num,   out=None):
+        """ Computes the approximate gradient for each selected function at :code:`x` given a `function_number` in {1,...,len(functions)}.
+
+        Parameters
+        ----------
+        x : DataContainer
+        out: return DataContainer, if `None` a new DataContainer is returned, default `None`.
+        function_num: `int` 
+            Between 1 and the number of functions in the list  
```

Comment:
Could be `0 and the number of functions in the list`

Reply:
Excellent spot!